### PR TITLE
Issue #4128 - test the decoding of OpenId Credentials

### PR DESCRIPTION
**Pull request by lachlan-roberts**
new tests for changes from PR #4129 for issue #4128

### CODE REVIEW

File: jetty-openid/src/main/java/org/eclipse/jetty/security/openid/CredentialsDecoder.java [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+import java.util.Map;
+
+import org.eclipse.jetty.util.ajax.JSON;
+import org.eclipse.jetty.util.log.Log;
+import org.eclipse.jetty.util.log.Logger;
+
+/**
+ * Used to decode the ID Token from the base64 encrypted JSON Web Token (JWT).
+ */
+public class CredentialsDecoder
```
Comment (travisspencer) [inline jetty-openid/src/main/java/org/eclipse/jetty/security/openid/CredentialsDecoder.java]: I'd suggest renaming this to `JwtDecoder`

File: jetty-openid/src/main/java/org/eclipse/jetty/security/openid/CredentialsDecoder.java [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+{
+    private static final Logger LOG = Log.getLogger(CredentialsDecoder.class);
+    private static final Base64.Decoder decoder = Base64.getUrlDecoder();
```
Comment (travisspencer) [inline jetty-openid/src/main/java/org/eclipse/jetty/security/openid/CredentialsDecoder.java]: is this thread-safe? I think that `Base64.getUrlDecoder` already manages its state to make subsequent calls faster than the initial one. Not 100% about that though.

File: jetty-openid/src/main/java/org/eclipse/jetty/security/openid/CredentialsDecoder.java [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+        return (Map)JSON.parse(jwtClaimString);
+    }
+
+    static byte[] padJWTSection(String unpaddedEncodedJwtSection)
+    {
+        int length = unpaddedEncodedJwtSection.length();
+        int remainder = length % 4;
+
+        if (remainder == 1)
+            throw new IllegalArgumentException("Not a valid Base64-encoded string");
```
Comment (travisspencer) [inline jetty-openid/src/main/java/org/eclipse/jetty/security/openid/CredentialsDecoder.java]: suggest adding back my comment:
`            // A valid base64-encoded string will never be have an odd number of characters.`

File: jetty-openid/src/test/java/org/eclipse/jetty/security/openid/CredentialsDecoderTest.java [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+import static org.junit.jupiter.api.Assertions.assertThrows;
+
+public class CredentialsDecoderTest
+{
+    public static Stream<Arguments> paddingExamples()
+    {
+        return Stream.of(
+            Arguments.of("XXXX", "XXXX"),
+            Arguments.of("XXX", "XXX="),
+            Arguments.of("XX", "XX==")
```
Comment (travisspencer) [inline jetty-openid/src/test/java/org/eclipse/jetty/security/openid/CredentialsDecoderTest.java]: would add `XXX=` and maybe `X=` to see what happens

File: jetty-openid/src/test/java/org/eclipse/jetty/security/openid/CredentialsDecoderTest.java [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.is;
+import static org.junit.jupiter.api.Assertions.assertThrows;
+
+public class CredentialsDecoderTest
+{
+    public static Stream<Arguments> paddingExamples()
+    {
+        return Stream.of(
+            Arguments.of("XXXX", "XXXX"),
```
Comment (travisspencer) [inline jetty-openid/src/test/java/org/eclipse/jetty/security/openid/CredentialsDecoderTest.java]: would add special chars as well like `X-X` and ensure that it comes back as `X-X=`.

File: jetty-openid/src/test/java/org/eclipse/jetty/security/openid/OpenIdAuthenticationTest.java [hunk 6]
```diff
@@ -149,8 +149,8 @@ public void testLoginLogout() throws Exception
         content = response.getContentAsString().split("\n");
         assertThat(content.length, is(3));
         assertThat(content[0], is("userId: 123456789"));
-        assertThat(content[1], is("name: FirstName LastName"));
-        assertThat(content[2], is("email: FirstName@fake-email.com"));
+        assertThat(content[1], is("name: Alice"));
+        assertThat(content[2], is("email: Alice@fake-email.com"));
```
Comment (travisspencer) [inline jetty-openid/src/test/java/org/eclipse/jetty/security/openid/OpenIdAuthenticationTest.java]: fake-email.com is an actual domain. Would change to example.com which is not possible to register.

File: jetty-openid/src/test/java/org/eclipse/jetty/security/openid/JwtEncoder.java [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+
+    public static String createIdToken(String provider, String clientId, String subject, String name, long expiry)
+    {
+        return "{" +
+            "\"iss\": \"" + provider + "\"," +
+            "\"sub\": \"" + subject + "\"," +
+            "\"aud\": \"" + clientId + "\"," +
+            "\"exp\": " + expiry + "," +
+            "\"name\": \"" + name + "\"," +
+            "\"email\": \"" + name + "@fake-email.com" + "\"" +
```
Comment (travisspencer) [inline jetty-openid/src/test/java/org/eclipse/jetty/security/openid/JwtEncoder.java]: would not use an actual domain `fake-email.com` but instead use `example.com`.

File: jetty-openid/src/main/java/org/eclipse/jetty/security/openid/CredentialsDecoder.java (hunk 3 above)
Comment (lachlan-roberts) [reply jetty-openid/src/main/java/org/eclipse/jetty/security/openid/CredentialsDecoder.java]: Not sure about this, what about the case with a remainder of 3, that is an odd number of characters.

File: jetty-openid/src/test/java/org/eclipse/jetty/security/openid/CredentialsDecoderTest.java (hunk 4 above)
Comment (lachlan-roberts) [reply jetty-openid/src/test/java/org/eclipse/jetty/security/openid/CredentialsDecoderTest.java]: How do you think it should behave with input `X=`. Right now it would still pad that to `X===`, but this actually an illegal base64 string length without the padding.

Maybe we should not pad any string which ends with `=`, or perhaps we should just throw illegal argument because it is already padded.

File: jetty-openid/src/main/java/org/eclipse/jetty/security/openid/CredentialsDecoder.java (hunk 2 above)
Comment (lachlan-roberts) [reply jetty-openid/src/main/java/org/eclipse/jetty/security/openid/CredentialsDecoder.java]: I think it is thread-safe because `Base64.getUrlDecoder()` returns you its static field `RFC4648_URLSAFE`
where `static final Decoder RFC4648_URLSAFE = new Decoder(true, false);`

but I have changed it to a local variable which is probably clearer

File: jetty-openid/src/main/java/org/eclipse/jetty/security/openid/CredentialsDecoder.java (hunk 3 above)
Comment (travisspencer) [reply jetty-openid/src/main/java/org/eclipse/jetty/security/openid/CredentialsDecoder.java]: touché! ⚔️ 

`// A valid base-64-encoded string will have a remainder of 0, 2 or 3. Never 1!`

I just think that this case is not obvious and we should help our future selves with some comment as a reminder. If you disagree, we can leave this comment out. If we add one though, you're *certainly* right that it should be correct 😄

File: jetty-openid/src/test/java/org/eclipse/jetty/security/openid/CredentialsDecoderTest.java (hunk 4 above)
Comment (travisspencer) [reply jetty-openid/src/test/java/org/eclipse/jetty/security/openid/CredentialsDecoderTest.java]: I could see a poorly implemented OP send back padded base64-encoded JWT sections. In this case, I think we should just try to use it as the OP gave and not add padding. Or, if the OP didn't pad enough, maybe we should add the required amount of padding. I'm just trying to make sure that we're as tolerant of what we receive while being as strict about what it is we send.

File: jetty-openid/src/main/java/org/eclipse/jetty/security/openid/CredentialsDecoder.java (hunk 3 above)
Comment (lachlan-roberts) [reply jetty-openid/src/main/java/org/eclipse/jetty/security/openid/CredentialsDecoder.java]: Yep I agree, have added this comment in.

File: jetty-openid/src/test/java/org/eclipse/jetty/security/openid/CredentialsDecoderTest.java (hunk 4 above)
Comment (lachlan-roberts) [reply jetty-openid/src/test/java/org/eclipse/jetty/security/openid/CredentialsDecoderTest.java]: Okay, I changed `padJWTSection()` so we no longer try to add padding if the string ended with `=`.

File: jetty-openid/src/test/java/org/eclipse/jetty/security/openid/JwtDecoderTest.java [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+            Arguments.of("XXX", "XXX="),
+            Arguments.of("XX", "XX=="),
+            Arguments.of("XXX=", "XXX="),
+            Arguments.of("X-X", "X-X="),
+            Arguments.of("@#", "@#=="),
+            Arguments.of("X=", "X="),
+            Arguments.of("XX=", "XX="),
+            Arguments.of("XX==", "XX=="),
+            Arguments.of("XXX=", "XXX="),
+            Arguments.of("", "")
```
Comment (joakime) [inline jetty-openid/src/test/java/org/eclipse/jetty/security/openid/JwtDecoderTest.java:48]: These don't seem like valid Base64.

File: jetty-openid/src/main/java/org/eclipse/jetty/security/openid/JwtDecoder.java [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+    public static Map<String, Object> decode(String jwt)
+    {
+        if (LOG.isDebugEnabled())
+            LOG.debug("decode {}", jwt);
+
+        String[] sections = jwt.split("\\.");
+        if (sections.length != 3)
+            throw new IllegalArgumentException("JWT does not contain 3 sections");
+
+        Base64.Decoder decoder = Base64.getUrlDecoder();
```
Comment (joakime) [inline jetty-openid/src/main/java/org/eclipse/jetty/security/openid/JwtDecoder.java:51]: You can initialize this once as a static field on the class.

File: jetty-openid/src/test/java/org/eclipse/jetty/security/openid/JwtDecoderTest.java (hunk 8 above)
Comment (joakime) [reply jetty-openid/src/test/java/org/eclipse/jetty/security/openid/JwtDecoderTest.java:48]: Can we see valid `.decode()` examples being tested?
This whole padding thing seems way off (there's so many modes in the java Base64, one of them has got to handle this without the strange `padJWTSection()` hack)..
The use of the a `UrlDecoder` is also suspect (to me).
Still don't see a real world scenario / testcase that demonstrates the need for this padding concept (vs just using the correct Base64 decoder setup/configuration from the get go)

Comment (lachlan-roberts) [reply jetty-openid/src/test/java/org/eclipse/jetty/security/openid/JwtDecoderTest.java:48]: `padJWTSection()` only adds the required padding to a string, it does not verify that the contents of the string are valid base64. I would have done more tests for `JwtDecoder.decode()` itself, but its hard to find example JWT strings without revealing personal information in them, which is why I was just generating a fake JWT.


As for why `padJWTSection` is necessary, is because the padding is manually removed when creating the JWS, see in [RFC7515](https://tools.ietf.org/html/rfc7515#section-2) it says:
```
Base64 encoding using the URL- and filename-safe character set
defined in Section 5 of RFC 4648 [RFC4648], with all trailing '='
characters omitted
```
and then gives an example of how to do the URL safe decoding and addition of the padding.
https://tools.ietf.org/html/rfc7515#appendix-C

File: jetty-openid/src/main/java/org/eclipse/jetty/security/openid/JwtDecoder.java (hunk 9 above)
Comment (lachlan-roberts) [reply jetty-openid/src/main/java/org/eclipse/jetty/security/openid/JwtDecoder.java:51]: If you make this a static field you need to know that it is shareable, but that is not immediately obvious without clicking through to the implementation.

**Comment by travisspencer** [comment]
@joakime , it would be great if Java provided a decoder/encoder that worked with JOSE, but, to the best of my knowledge, it doesn't. That knowledge has been gained by built an entire OpenID Connect server in Java writing similar Java modules to this for use in Web apps and APIs.

The "real-world" examples may not exist in the test, but what @lachlan-roberts has added are the quintessentials plus some more exotic corner cases that we should not expect to receive but can now handle in case we do. The reason for this is that the decoding is done on _parts_ of the JWT. a JWS (a signed JWT) is made up of 3 parts separated by a dot. This special decoding has to be done on each of these 3 parts. The "hacky" `padJWTSection` method (which is similar to the example in the JOSE RFC, so not sure how "hacky" it is BTW), is just to decode each of these parts of the JWS. For this reason, the real world input would need to execute more than just the code that's currently under test.

Hope this clarifies and relieves any anxiety you might have about this test or this new code.

**Comment by joakime** [comment]
I have nothing to add to this PR.
At this point I'm neutral on it.
My only concerns were with the Base64 behaviors, not the OpenID behaviors.
I'm withdrawing review request entirely.

**Comment by lachlan-roberts** [comment]
@travisspencer could we use your example from #4128 in a test as a real world example which demonstrates this issue?

File: jetty-openid/src/main/java/org/eclipse/jetty/security/openid/JwtDecoder.java [hunk 10]
```diff
@@ -53,7 +53,10 @@
         String jwtClaimString = new String(decoder.decode(padJWTSection(sections[1])), StandardCharsets.UTF_8);
         String jwtSignature = sections[2];
 
-        Map<String, Object> jwtHeader = (Map)JSON.parse(jwtHeaderString);
+        Object parsedJwtHeader = JSON.parse(jwtHeaderString);
+        if (!(parsedJwtHeader instanceof Map))
+            throw new IllegalStateException("Invalid JWT header");
+        Map<String, Object> jwtHeader = (Map)parsedJwtHeader;
```
Comment (travisspencer) [inline jetty-openid/src/main/java/org/eclipse/jetty/security/openid/JwtDecoder.java:59]: This is not the safest of casts. I'd check that the keys are strings using code like we've done in previous PRs.

File: jetty-openid/src/main/java/org/eclipse/jetty/security/openid/JwtDecoder.java [hunk 11]
```diff
@@ -63,7 +66,10 @@ and the Token Endpoint (which it is in this flow), the TLS server validation
         if (LOG.isDebugEnabled())
             LOG.debug("JWT signature not validated {}", jwtSignature);
 
-        return (Map)JSON.parse(jwtClaimString);
+        Object parsedClaims = JSON.parse(jwtClaimString);
+        if (!(parsedClaims instanceof Map))
+            throw new IllegalStateException("Could not decode JSON for JWT claims.");
+        return (Map)parsedClaims;
```
Comment (travisspencer) [inline jetty-openid/src/main/java/org/eclipse/jetty/security/openid/JwtDecoder.java:72]: This method returns a typed map. These types should be checked as well before returning.

**Comment by travisspencer** [comment]
Yes, @lachlan-roberts . That JWT should work. You can decode it using oauth.tools, BTW. There, you can see that the `purpose` claim is `id`, so it's an ID token like the one that an OP will issue to Jetty.

File: jetty-openid/src/main/java/org/eclipse/jetty/security/openid/JwtDecoder.java (hunk 10 above)
Comment (lachlan-roberts) [reply jetty-openid/src/main/java/org/eclipse/jetty/security/openid/JwtDecoder.java:59]: Would a Map with keys that are not strings even be valid JSON. If our JSON parser is capable of returning invalid JSON then we should fix that in the parser rather than testing the return type is valid everywhere it is used.

Comment (travisspencer) [reply jetty-openid/src/main/java/org/eclipse/jetty/security/openid/JwtDecoder.java:59]: That's a fair point, and do not oppose this change given that argument. However, I still would do the conversion at run-time to a `Map<String, Object>` and not cast it to that. We're not in a high throughput case here. We can tolerate 250ms for the entire code flow + login (say 6 request/responses) in most cases I've seen. So, I'd err on the side of clean code that relies on the compiler for type checking. As I said though, your argument is hard to dispute.

File: jetty-openid/src/test/java/org/eclipse/jetty/security/openid/JwtEncoder.java [hunk 12]
```diff
@@ -0,0 +1,52 @@
...
+//
+//  You may elect to redistribute this code under either of these licenses.
+//  ========================================================================
+//
+
+package org.eclipse.jetty.security.openid;
+
+import java.util.Base64;
+
+public class JwtEncoder
```
Comment (janbartel) [inline jetty-openid/src/test/java/org/eclipse/jetty/security/openid/JwtEncoder.java:26]: Comment on the purpose of the class.

File: jetty-openid/src/test/java/org/eclipse/jetty/security/openid/JwtEncoder.java [hunk 13]
```diff
@@ -0,0 +1,52 @@
...
+    }
+
+    private static String stripPadding(String paddedBase64)
+    {
+        return paddedBase64.split("=")[0];
+    }
+
+    public static String createIdToken(String provider, String clientId, String subject, String name, long expiry)
+    {
+        return "{" +
```
Comment (janbartel) [inline jetty-openid/src/test/java/org/eclipse/jetty/security/openid/JwtEncoder.java:49]: Use StringBuilder?

Also, a comment on the expected format that this method generates might be useful, or a link to a document somewhere that describes it.

File: jetty-openid/src/test/java/org/eclipse/jetty/security/openid/JwtEncoder.java (hunk 12 above)
Comment (lachlan-roberts) [reply jetty-openid/src/test/java/org/eclipse/jetty/security/openid/JwtEncoder.java:26]: This class is not production code, this class is used only for testing.
I will a javadoc comment with an explanation of what it does.

File: jetty-openid/src/test/java/org/eclipse/jetty/security/openid/JwtEncoder.java (hunk 13 above)
Comment (lachlan-roberts) [reply jetty-openid/src/test/java/org/eclipse/jetty/security/openid/JwtEncoder.java:49]: I just tried this but intellij is suggesting I use string concatenation instead of StringBuilder for better code readability, apparently using StringBuilder in this case is not making the code more efficient.

**Comment by travisspencer** [comment]
@janbartel , I've looked again. The handling of the JWTs seems good to me.

**Comment by lachlan-roberts** [comment]
Thanks @travisspencer!
I have also added the test with the JWT example from #4128 in commit 67ec3d2e87b74f632c18ba5df0cd639ea7789671.
I will go ahead and merge.
